### PR TITLE
[@types/libav-core] Init libav-core types

### PR DESCRIPTION
**Pull request by jeromewu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

This is a special type that is meant to be used by multiple npm packages, so disabled `npm-naming`.

### CODE REVIEW

**Comment by peterblazejewicz** [comment]
what's the package name here? Is it `@ffmpeg/libav-core`? Can you link published package? (unpkg/npm)

**Comment by jeromewu** [comment]
@peterblazejewicz 

Thanks for reviewing this PR.

I just published two packages that uses these type at the moment:

https://www.npmjs.com/package/@ffmpeg/libav-core
https://www.npmjs.com/package/@ffmpeg/libav-core-mt

And both use the same d.ts here: https://github.com/ffmpegwasm/libav.wasm/blob/main/src/libav-core.d.ts

So the idea here is that I would like to use the same .d.ts for these two packages as their APIs are the same, but `@ffmpeg/libav-core` is single thread version and `@ffmpeg/libav-core-mt` is multi thread version which requires `SharedArrayBuffer`.

In the future, if any users built their own libav-core, it is possible that they also reuse `@types/libav-core` here directly.

File: types/libav-core/tslint.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "@definitelytyped/dtslint/dt.json",
+  "rules": {
+    "npm-naming": false
```
Comment (weswigham) [inline types/libav-core/tslint.json:4]: Please don't disable this lint rule - in this case, I think you've got the package name wrong. If this is for this package: https://www.npmjs.com/package/@ffmpeg/libav-core

You should write types for in a `ffmpeg__libav-core` folder and write `@ffmpeg/libav-core` as the package name in the header.

**Review by weswigham — CHANGES_REQUESTED** [review]

**Comment by weswigham** [comment]
> So the idea here is that I would like to use the same .d.ts for these two packages as their APIs are the same

Just have one package reexport the other, here on DT. You should still write two packages, though.

File: types/libav-core/tslint.json (hunk 1 above)
Comment (jeromewu) [reply types/libav-core/tslint.json]: Thanks for the suggestion, fixed with a new commit.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ffmpeg__libav-core ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ffmpeg__libav-core@latest/))
was missing the following properties:
1. Module
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9a53a89824cb43b86d04ef9e62c4d828676be452
</p>

File: types/ffmpeg__libav-core/ffmpeg__libav-core-tests.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import createLibavCore from 'ffmpeg__libav-core';
```
Comment (peterblazejewicz) [inline types/ffmpeg__libav-core/ffmpeg__libav-core-tests.ts:1]:
```suggestion
import createLibavCore from '@ffmpeg/libav-core';
```

File: types/ffmpeg__libav-core/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "esModuleInterop": true,
+        "baseUrl": "../",
+        "typeRoots": [
+            "../"
+        ],
```
Comment (peterblazejewicz) [inline types/ffmpeg__libav-core/tsconfig.json:15]: you would also have to add paths translation here for scoped module to be properly resolved from outside:
https://github.com/DefinitelyTyped/DefinitelyTyped#what-about-scoped-packages

File: types/ffmpeg__libav-core/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "esModuleInterop": true,
```
Comment (peterblazejewicz) [inline types/ffmpeg__libav-core/tsconfig.json:11]:
```suggestion
```
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports
this is not used on DT, unless for very specific reason, I'll comment on your imports.
it's used when consuming DT, e.g. with defaults in tsconfig

File: types/ffmpeg__libav-core/ffmpeg__libav-core-tests.ts (hunk 2 above)
Comment (jeromewu) [reply types/ffmpeg__libav-core/ffmpeg__libav-core-tests.ts]: Fixed, thanks!

File: types/ffmpeg__libav-core/tsconfig.json (hunk 3 above)
Comment (jeromewu) [reply types/ffmpeg__libav-core/tsconfig.json:14]: Added path to tsconfig.json, thanks!

File: types/ffmpeg__libav-core/tsconfig.json (hunk 4 above)
Comment (jeromewu) [reply types/ffmpeg__libav-core/tsconfig.json]: Reomved, thanks!

**Comment by jeromewu** [comment]
@peterblazejewicz Thanks for the reviews! I learnt a lot from your suggestions. :)

File: types/ffmpeg__libav-core/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Type definitions for @ffmpeg/libav-core 0.0
```
Comment (peterblazejewicz) [inline types/ffmpeg__libav-core/index.d.ts:1]: so the distributed package seems to contain types already, what's the relation of those o this PR?
https://unpkg.com/browse/@ffmpeg/libav-core@0.0.1-alpha.58/package.json
https://unpkg.com/browse/@ffmpeg/libav-core@0.0.1-alpha.58/dist/libav-core.d.ts

Comment (jeromewu) [reply types/ffmpeg__libav-core/index.d.ts:1]: Removed types in this package, thanks for the reminder!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@jeromewu thanks!

**Comment by jeromewu** [comment]
Ready to merge

**Pull request merged**